### DISCUSSION TITLE
Feature proposal : observations.org upload

### DISCUSSION THREAD

**Discussion by alexbelgium** [Nachtzuster/BirdNET-Pi#130] [Ideas]
Thread used to aggregate ideas on upload to observations.org

**Comment by alexbelgium** [comment]


**Reply by Hiemstra87** [reply]
Nice for 'hunting' species. 
For me personally this is not the end device for recording observations (I use Observation.org for that), so for me an upload button per recording would be nicer (as the confirmation takes place there, and in most cases are moderated as well). But that might be a different story, observation.org is pretty Euro oriented and i'm not sure they have a way for integrating easily.

**Reply by alexbelgium** [reply]
Manual recording observations would be complex... What would be better is for observations.org to provide an API for that, as birdweather.com is doing. This is not very likely though :-)

Edit : seems feasible actually https://github.com/mcguirepr89/BirdNET-Pi/discussions/1082#discussion-5941748

**Reply by Hiemstra87** [reply]
They do reply very quick (about an API request):
Hi Bouke,

Thanks for reaching out. We are definitely interested, and we even think the technology already exists. We have similar functionality for the Belgian monitoring software (Meetnetten.be), where users can

Authenticate themselves with their Observation.org account

Send project observations to their personal Observation.org account

We have an API available for this, you can find the documentation at https://observation.org/api/v1/docs/

For API access, we need to have a user account on Observation.org that is responsible for the API use.

I suggest we do have a short discussion about functional issues, such as observation type, adding media to the observation, rewriting observations after changes on either platform, et cetera. And about the way we explain this to users, and perhaps even do a little promotional activity if we get this show on the road.

Also, if you want, we can speak Dutch, and we have BirdNET-PI users in our own dev team :-)

Looking forward to see what this can bring for BirdNET-PI and Observation!

All the best,
Dylan from Observation.org

-----------------
My question would be, is this something we want to add in this fork, or do we have to look upstream?

**Reply by alexbelgium** [reply]
Wow incredible!! Well then I can look at it but no promises that I'll have time as it is soon holidays. Observations.org is very popular indeed so this should probably be a much wanted feature

Edit : BTW there is no real upstream ; this fork is the upstream ;)

**Reply by Hiemstra87** [reply]
Would be real nice indeed. I personally know at least 3 B-PI users who would use this function on a daily basis. And of course holidays are important too. I'm not a real coder, but i will see were i can chip in, as this is highly interesting to me.

**Reply by alexbelgium** [reply]
Here is the branch where I'll start some ideas : https://github.com/alexbelgium/BirdNET-Pi/tree/observations_org - however Nachzuster is a much better coder than me so once I'll have something showable I'll give it to him so he can see how we can integrate it in a better way :-)

**Reply by alexbelgium** [reply]
Mmh I don't understand in their API if the client_id would be the same `curl -X POST https://observation.org/api/v1/oauth2/token/ \
     -d "client_id=<client_id>&grant_type=password&email=<email>&password=<passsword>"` from https://observation.org/api/v1/docs/authentication-oauth2.md

**Reply by Hiemstra87** [reply]
When we have a bit of understanding of the API i can set something up to discuss with the observation.org devs

**Reply by alexbelgium** [reply]
easiest would be if they came to the github PR once we have something working ;-)
I'm no dev either but tinkerer enough to put in place something. I'm still working on code logic, then once it's done we can think about pulling them in the PR https://github.com/Nachtzuster/BirdNET-Pi/pull/142

**Comment by YvedD** [comment]
Hi !
I'm no coder, but i use observations on a daily basis.
Just thinking about "automated" upload ... would that not be tricky with false positives, where manual upload would leave room to check the recordings and then decide if it has to be uploaded?
Just thinking along.

Otherwise, i would be a great add-on for this Birdnet-Pi fork.

Grtz,
Yves

**Reply by alexbelgium** [reply]
Yes, probably much better. And easier to implement ;-) let's start with manual and then we'll see if we want to move it further. Already it will be a huge effort. Their API is actually excellent and can also upload the sound. My main issue for the moment is understanding how to authentificate in the API ;-)

**Reply by Hiemstra87** [reply]
Also, auto upload might be something Observations.org might not like, seen as how many uploads that might cause. Also they tend to moderate all observations (unless very common species). Their data is used in many research papers, so the data must remain somewhat reliable. 
These are indeed some of the 'rules' we will discuss with them as well.

**Reply by alexbelgium** [reply]
Fully agree ; auto upload would not make initial sense. I've started a script in a branch, I'll do a PR to track work there : https://github.com/Nachtzuster/BirdNET-Pi/pull/142

**Comment by alexbelgium** [comment]
It's progressing - as far as I can go probably without a client_id to fetch a token. 
What we would need as info from observation.org devs at this point :
- We need a client_ID (so I can try the code to fetch tokens)
- Can the same client_ID be used for all birdnet-pi (knowing each user will use his own password/username)
- Can it be distributed on github or is there a way to obscurize it (we can always think of something)

**Reply by Hiemstra87** [reply]
I'll send an invite to them tonight to get on this Git to discuss these questions and the way forward.

**Reply by alexbelgium** [reply]
Just for info I'll be gone for 2 weeks and likely not available. Working : new settings appear in the config page (username and password only when the site is defined), the button to upload appears on individual observations (only if an upload site is defined). When clicked it gathers all information for the observation in a json format including the sound file to upload then should upload it to observation.org. Once uploaded, it stores the uuid in a txt file along with the upload site and the filename.

The only missing part that I haven't tested in all that is the upload part, and the writing of the uuid to the file

**Reply by YvedD** [reply]
I'm following this with interest.
I'm from Belgium too, if i can provide any help ...shout.
I use BirdNET-Pi for my NFC's.

**Reply by alexbelgium** [reply]
Btw this approach can perfectly be adapted to other sites such as inaturalist with very minimal updates. The code is set in a general manner to allow this

**Reply by alexbelgium** [reply]
> I'm following this with mush interest. I'm from Belgium too, if i can provide any help ...shout. I use BirdNET-Pi for my NFC's.

Thanks! Don't hesitate to double check my code for optimization :) I'm not a true dev but work ;) the key thing I haven't verified is the upload function although I've written an early draft

**Reply by Hiemstra87** [reply]
I'll set a test rig up tomorrow as well so I can test with the guys and girls from Observation.org. 
@alexbelgium 
Thanks for the work, and have a great vacation!

**Reply by alexbelgium** [reply]
@Hiemstra87 Forgot one thing : the push api script requires a "species id" field, but I don't know how to convert the Scientific Name into a observation.org species id. It seems a dictionary should be needed ?

**Reply by Hiemstra87** [reply]
Np, i'll discuss option with them on this. If this really needs a dict we could probably get a conversion table from them. But let's see what the requirements are for species identification.

**Reply by alexbelgium** [reply]
Thanks - all done in theory ; only remaining thing would be to get the "species id" conversion and a "client id" to be able to test an upload. I've added a notes field so that the user can append a custom note when he is posting to observation.org. I've also prepared the set for inaturalist (much more used in anglosaxon countries) but that will be a next step

**Reply by alexbelgium** [reply]
Dictionary is not needed anymore I've got the species id from the api

**Reply by Hiemstra87** [reply]
Very productive you've been. Much appreciate

**Comment by Hiemstra87** [comment]
Sorry for the late reaction here, but here is a reaction from the people at Observation.org:

We can provide you with the client ID. You absolutely must not publish this, so it will need to be recorded as a secret on GH. The client ID is our way of not giving unlimited access to resources (computing capacity).

I have briefly read the discussion. It seems like a good idea to have a short call about functionality and technology.

We also want any observations from BirdNET-PI to be categorized correctly in our system (as an automated observation, not as an observation made by a person in the field).

Please understand that the Dutch holiday period is approaching. In July and August, we are a bit slower than usual.

**Reply by Hiemstra87** [reply]
I think it would be beneficial to have a call setup between us and them to get alignment on the way we set up this cooperation with them (what and how we share info to their platform).

To manage expectations in this holiday season: I expect less availability from them (and from our side) between now and the end of august. 

Let me come back to you all on the how and when. You can always DM me your earliest available timeslot (or two) so i can start planning a call (if that is what you want to agree to). From my end i'd really like to discuss how our observations will be registered on their platform (manual vs automated).

@alexbelgium Thanks for the pre work already done, great ground work for further integrations as well!

I know these times are for holidays as well, and this project is all done voluntarily, so no pressure on this 😉

**Reply by alexbelgium** [reply]
Hi, this is understandable. I'm also on holidays. The system is defined in a way that we can also use ebird or inaturalist API instead

**Reply by alexbelgium** [reply]
My thoughts : observations should be recorded as manual, as the user needs to manually verify then submit each observation. In this case, the birdnet detection is more an help for the user but should not be the only way of confirmation

We cannot keep the client id as a github secret as the systems are not ran in github pages but in standalone systems such as raspios. We could imagine another system such as using php minify to make it harder to be readable though. Or compile a mini binary to publish the observation? Here is a page I found with several options https://www.baeldung.com/linux/bash-obfuscate-script

**Reply by alexbelgium** [reply]
> I think it would be beneficial to have a call setup between us and them to get alignment on the way we set up this cooperation with them (what and how we share info to their platform).

I was looking for a way to start secure personal communication. Are you using telegram? I've created a group that you can join with this link. https://t.me/+gmfSptJ7hrI5MGFk This should provide a secure and private manner of communicating and setting up the API discussion that you were mentioning. I would propose that only people with a direct added value to this specific piece of work are included - as you may know communication channels grow at the rate of n*(n-1)/2 where n is the number of participants so we should keep the number as low as possible.

**Reply by Hiemstra87** [reply]
> My thoughts : observations should be recorded as manual, as the user needs to manually verify then submit each observation. In this case, the birdnet detection is more an help for the user but should not be the only way of confirmation
> 
> We cannot keep the client id as a github secret as the systems are not ran in github pages but in standalone systems such as raspios. We could imagine another system such as using php minify to make it harder to be readable though. Or compile a mini binary to publish the observation? Here is a page I found with several options https://www.baeldung.com/linux/bash-obfuscate-script

Could we still use Github and access that from the app to retrieve the ID & secret? 
I read about some options Github can provide:

- Environment Variables: Store your API keys and other sensitive information in environment variables instead of hardcoding them in your source code. You can then access these environment variables in your code at runtime.

- .env Files: Use a .env file to store your API keys and other sensitive information. The .env file should not be committed to your repository. You can use a library like python-dotenv in Python or dotenv in Node.js to load the variables from the .env file.

- Git Ignore: Add your .env file to your .gitignore file so that it is not tracked by Git and not pushed to your GitHub repository.

- Secrets Management: GitHub provides a feature called GitHub Secrets that allows you to store sensitive information securely. You can store your API keys as secrets in your GitHub repository settings and then access them in your GitHub Actions workflows or other parts of your project.

Other options might be:
- Configuration Files: Store your API keys in a separate configuration file that is not tracked by Git. You can provide a template configuration file with placeholders for the keys and instruct users to create their own local configuration file.

- Encryption: Encrypt your API keys before storing them in your repository and decrypt them at runtime. Tools like git-crypt can help you encrypt sensitive files in your repository.

We might discuss these options with Observation as well, to see what they are comfortable with.

**Reply by Hiemstra87** [reply]
> > I think it would be beneficial to have a call setup between us and them to get alignment on the way we set up this cooperation with them (what and how we share info to their platform).
> 
> I was looking for a way to start secure personal communication. Are you using telegram? I've created a group that you can join with this link. https://t.me/+gmfSptJ7hrI5MGFk This should provide a secure and private manner of communicating and setting up the API discussion that you were mentioning. I would propose that only people with a direct added value to this specific piece of work are included - as you may know communication channels grow at the rate of n*(n-1)/2 where n is the number of participants so we should keep the number as low as possible.

Agree, just you and Alexandre should be enough, maybe me but i'm not sure if i'm a lot of help in coding this.

I can ask them if the can join the telegram group when we get confirmation from Alexandre that he can join as well.

**Reply by alexbelgium** [reply]
Who is "you" vs Alexandre ? If "you" refer to me then I am Alexandre :-)

**Reply by Hiemstra87** [reply]
Haha, i thought i was reacting to Nachtzuster...

**Reply by alexbelgium** [reply]
No problem for me to talk to the folks of observation.org through the telegram channel if they agree, or any other communication medium - and for you it all depends on how you see your level of involvement and where you can support!

**Reply by Hiemstra87** [reply]
> No problem for me to talk to the folks of observation.org through the telegram channel if they agree, or any other communication medium - and for you it all depends on how you see your level of involvement and where you can support!

I invited them through the url, and entered the channel myself, to see where I might be useful. Thanks!

**Reply by YvedD** [reply]
i think that the users MUST apply for a personal API-Key, and that they have to put it in place themselves after receiving it from observation.com.
Why am i thinking of this :
It's the responsibility of the users to ask for a API key if they want their observations to be uploaded to that platform, as it is their responsibility to check the observation on false positives etc.
I have asked different API Key's for online weather forecasts, and the good use of these keys is up to me as a user of the platform, no?
That way, one can prevent "miss-use" of a general API-key (imagine someone hacking a genral key and uoloading 1.000.000 records of a passer domesticus !!!),beside that - for "on-air" public systems i should be encrypted anyway ... or am i drifting too far off? 

Grtz,
Yves

**Reply by alexbelgium** [reply]
Hi, thanks for your thoughts! Actually in the way the API works the upload will already be linked to the username and password of the observation.org user to this aim (those 2 infos will have to be encoded in the config of BirdNET pi). However, the client id is more a specific code that the obs.org admin use to identify which system is used to encoding, and they want this hidden

Br

**Reply by Hiemstra87** [reply]
Reply from Observation:
Due to the holidays they are limited in respons time. Expect a reply from them in the first half of august.

**Reply by alexbelgium** [reply]
Any movement from their side ?

Critical element will be the complexity of managing the client ID as observations would be sent from the local systems, and by definition in an open-code app everything is visible. Perhaps the php code could query the client ID from github

Or the easiest could be that each user requests a specific client ID, using `grant_type=client_credentials`

**Reply by roelvanderkraan** [reply]
Another way of handling secrets in open code is to have an intermediary worker (like a Cloudflare worker) which is a simple backend that handles the communication with observation.org. That way you keep the key out of the open source code and can limit interactions with the API to just the functions that are required for this functionality. Downside is that it adds a dependency on an outside service.

**Reply by Hiemstra87** [reply]
Just asked again to join the comm channel you set up @alexbelgium 
I think this might have escaped their attention (and mine as well, sorry for that)

**Comment by alexbelgium** [comment]
As a note : it is possible to prefill an upload with `https://observation.org/fieldwork/observations/create/?species=111&date=2024-05-12&time=11:11&lat=11&lng=12&notes=birdnet-pi&method=588&advanced=on` for example ; the only missing thing is how to upload ! Otherwise we might imagine a new window that opens to upload an observation

Edit : method adapted to "automated recorder"

**Reply by Hiemstra87** [reply]
Great idea, I'm not sure how uploading would work that way without a download and upload from the user self though.

**Reply by alexbelgium** [reply]
What I could do is have a script to open the observation page prefilled, and initiate a download of the mp3. The user would therefore then manually upload the mp3. 

It would significantly make the workflow easier, while making it long enough that people would not be tempted to upload tons of misidentified noisy recordings online ;)

**Reply by Hiemstra87** [reply]
I could live with that idea, although it would require a little bit of instruction for users. 
And until we hear again from out friends at Observation.org, it might be the only way.

Would recommend to pre fill the method to sound recorder. Might be good to mention when we do have contact with them, they could add Birdnet-PI as a method there. Would make it easy to filter

**Reply by alexbelgium** [reply]
I would more say "automated recognition" in the sense that birdnet-pi so one declination of a wider field of AI based recognition 

But the key idea is that it should be user validated based on the sound : so the method "heard" (or sound recorder) is indeed correct

**Reply by Hiemstra87** [reply]
True AR should be the best description. 

I must admit, sometimes when i'm in doubt on a sound I upload it as well (marked as uncertain) to let the moderators make a final judgement.
